components/brave_shields/browser/ad_block_subscription_source_provider.cc

```diff
@@ -0,0 +1,43 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_shields/browser/ad_block_subscription_source_provider.h"
+
+#include <utility>
+
+#include "base/logging.h"
+#include "base/task/thread_pool.h"
+#include "brave/components/brave_shields/common/pref_names.h"
+#include "components/prefs/pref_service.h"
+#include "content/public/browser/browser_task_traits.h"
+#include "content/public/browser/browser_thread.h"
+
+namespace brave_shields {
+
+AdBlockSubscriptionSourceProvider::AdBlockSubscriptionSourceProvider(
+    PrefService* local_state,
+    base::FilePath list_file)
+    : list_file_(list_file) {}
+
+AdBlockSubscriptionSourceProvider::~AdBlockSubscriptionSourceProvider() {}
+
+void AdBlockSubscriptionSourceProvider::ReloadListFromDisk() {
```

Comment:
I don't understand why we have two methods that do the same thing with different callbacks. The caller shouldn't hold a reference to an `AdBlockSubscriptionSourceProvider`, it only cares that it's a source provider and subclasses shouldn't add public methods like this

Reply:
I think the only exception would be the custom list provider because that needs an interface to update it so maybe SourceProvider and EditableSourceProvider?

Reply:
or SourceProvider, DATSourceProvider and PrefSourceProvider probably make more sense as the base classes

Reply:
where PrefsSourceProvider -> EditableSourceProvider -> SourceProvider and DATSourceProvider -> SourceProvider

Reply:
actually I take that back, the SourceProvider shouldn't be responsible for updating the list and should not be editable, it should only be responsible for loading the data when it changes. Because this is a kind of special case where the list is updated from the UI and not loaded from the network, the service should be responsible for updating the pref and then source provider should be a pref observer

Reply:
dat file source providers listen for component updates and pref source providers listen for pref changes, but source providers are not responsible for updating the data, they are only responsible for loading it and notifying the services when it changes